lib/HLSL/HLModule.cpp

```diff
@@ -231,17 +231,20 @@ void HLModule::RemoveFunction(llvm::Function *F) {
 namespace {
 template <typename TResource>
 bool RemoveResource(std::vector<std::unique_ptr<TResource>> &vec,
-                    GlobalVariable *pVariable, bool keepAllocated) {
+                    GlobalVariable *pVariable, bool keepAllocated,
+                    UnusedResourceBinding unusedResourceBinding) {
   for (auto p = vec.begin(), e = vec.end(); p != e; ++p) {
     if ((*p)->GetGlobalSymbol() != pVariable)
       continue;
 
-    if (keepAllocated && (*p)->IsAllocated()) {
+    if ((keepAllocated && (*p)->IsAllocated()) ||
+        unusedResourceBinding == UnusedResourceBinding::ReserveAll) {
```

Comment:
I think we could just unconditionally set the resource symbol to undef without removing it from the list, then rely on the `DxilCondenseResources` pass to clean up unused resources later as necessary. This simplifies reasoning about what code impacts this resource list too.

Reply:
I think the logic looks correct. What sort of test failure are you seeing?